designs/testing-design.md

```diff
@@ -0,0 +1,241 @@
+     self.a = a
+	 self.b = b
+  def concatenate(self)-> str:
+     return self.a + self.b
+  def RAfirst(self) -> str:
+     return self.a[0]
+  def RBFirst(self) -> str:
+     return self.b[0]
+  def LengthA(self) -> int:
+     return len(self.a)
+  def LengthB(self) -> int:
+     return len(self.b)
+A : Start = None
+A = Start()
+A.set("First","Second")
+print(A.concatenate())
+print(A.RAfirst())
+print(A.RBFirst())
+print(A.LengthA + A.LengthB)
+```
+Expected output:
+```
+FirstSecond
+F
+S
+11
+```
+#### Test Case 5 (Dictionary , Loops , Sets , Multiple Unpacking)
+```
+def Checking(n : int)-> Dict[int,int],Set[int]:
+  A = Dict[int,int] = None
+  B = Set[int] = None
+  i :int = 0
+  c : int = 0
+  i = -n
+  while i <= n:
+    c = pow(i,2)
+    A[i] = c
+    B.add(c)
+  return A,B
+a : Dict[int , int] = None
+b : Set[int] = None
+a,b = Checking(3)
+print(a)
+print(b)
+```
+Expected output:
+```
+{-3:9,-2:4,1:1,0:0,1:1,2:4,3:9}
+(0,1,4,9)
+```
+#### Test Case 6 (Inheritance+Polymorphism)
+```
+class Number(object):
+    x: int = 5
+    def getVal(self : Number) -> int:
+        return self.x
+class AddOne(Number):
+    def __init__(self: AddOne):
+        self.x = 0
+    def getVal(self: AddOne) -> int:
+        return self.x+1
+class AddTwo(Number):
+    def __init__(self: AddTwo):
+        self.x = 0
+    def getVal(self: AddTwo) -> int:
+        return self.x+2
+
+num1 : Number = None
+num2 : Number = None
+num1 = AddOne()
+print(num1.getVal())
+num2 = AddTwo()
+print(num2.getVal())
+```
+Expected output:
+```
+1
+2
+```
+#### Test Case 7 (Closures+For Loops+List)
+```
+def print_msg(val: int) -> Callable[[], None]:
+    def printDouble():
+        print(val*2)
+    return printDouble
+
+iterations: [int] = [1, 2, 3, 4]
+for i in iterations:
+    func: Callable[[], None] = print_msg(i)
+    func()
+```
+Expected output:
+```
+2
+4
+6
+8
+```
+
+#### Test Case 8 (Destructuring Assignment + I/O)
+```
+f1 : File, f2: File = None, None
+f1 = open('output1', 'wb')
+f2 = open('output2', 'wb')
+
+f1.write(4)
+f1.write(5)
+f1.write(5)
+f1.write(4)
+
+f1.close()
+f2.close()
+```
+Expected output:
+```
+# The test should result in two new files `output1` and `output2`
+# with the following content:
+output1: \x00000004\x00000005
+output2: \x00000005\x00000004
+```
+#### Test Case 9 (Closures+Iterators+List)
+```python=
+def print_msg(arr: [int]) -> Callable[[], None]:
+    def printFirstTwo():
+        itr = iter(arr)
+        print(next(itr))
+        print(next(itr))
+    return printFirstTwo
+printTwo: Callable[[], None] = print_msg([3, 4, 5])
+printTwo()
+```
+Expected output:
+```
+3
+4
+```
+#### Test Case 10 (Anonymous function + String)
+```
+isEvenLength: Callable[[string], bool] = lambda str:string: len(str)%2==0 
+print(isEvenLength("abc"))
+print(isEvenLength("abcd"))
+
+```
+Expected output:
+```
+False
+True
+```
+
+## Changes we want to make
+#### changes in tests/
+- Add more integrated test cases
```

Comment:
These are great tests!

I would suggest some tests that are a _little_ simpler than these, that focus on combining just, say, two features.

For example, just using all of the different kinds of new values as function arguments and class fields is a good way to make sure their memory allocation strategy/value representation is working well.

Or just doing destructuring assignment on the result of a comprehension (or a for loop on the result of a comprehension) will make sure those are tested together.

So think of 2-feature combinations.

All the ones you have are excellent and important, too!

Reply:
Got it. We will first focus on combining two features tests! Thank you!